generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,347 @@
+// The result of parsing a `google.api.http` annotation.
+//
+// A `google.api.http` annotation describes how to convert gRPC RPCs to HTTP
+// URLs. The description uses a "path template", showing what portions of the
+// URL path are replaced with values from the gRPC request message.
+//
+// These path templates follow a specific grammar. The grammar is defined by:
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = "*" | "**" | LITERAL | Variable ;
+//	Variable = "{" FieldPath [ "=" Segments ] "}" ;
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+//
+// The specific notation is not defined, but it seems inspired by
+// [Backus-Naur Form]. In this notation, `{ ... }` allows repetition.
+//
+// The documentation goes on to say:
+//
+//	A variable template must not contain other variables.
+//
+// So the grammar is better defined by:
+//
+//	Template = "/" Segments [ Verb ] ;
+//	Segments = Segment { "/" Segment } ;
+//	Segment  = Variable | PlainSegment;
+//	PlainSegment  = "*" | "**" | LITERAL ;
+//	Variable = "{" FieldPath [ "=" PlainSegments ] "}" ;
+//	PlainSegments = PlainSegment { "/" PlainSegment };
+//	FieldPath = IDENT { "." IDENT } ;
+//	Verb     = ":" LITERAL ;
+//
+// Neither "IDENT" nor "LITERAL" are defined. From context we can infer that
+// IDENT must be a valid proto3 identifier, so matching the regular expression
+// `[A-Za-z][A-Za-z0-9_]*`. Likewise, we can infer that LITERAL must be a path
+// segment in a URL. [RFC 3986] provides a definition for these, which we
+// summarize as:
+//
+// LITERAL     = pchar { pchar }
+// pchar       = unreserved | pct-encoded | sub-delims | ":" | "@"
+// unreserved  = ALPHA | DIGIT | "-" | "." | "_" | "~"
+// pct-encoded = "%" HEXDIG HEXDIG
+// sub-delims  = "!" | "$" | "&" | "'" | "(" | ")"
+//
+//	| "*" | "+" | "," | ";" | "="
+//
+// ALPHA       = [A-Za-z]
+// DIGIT       = [0-9]
+// HEXDIG      = [0-9A-Fa-f]
+//
+// [RFC 3986]: https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
+// [Backus-Naur Form]: https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+
+func Parse(pathTemplate string) (*PathTemplate, error) {
+	return parsePathTemplate(pathTemplate)
+}
+
+// PathTemplate represents the structure in Go.
+type PathTemplate struct {
+	Segments []*Segment
+	Verb     *Literal
+}
+
+// Match represents a single '*' match.
+type Match struct{}
+
+// MatchRecursive represents a '**' match.
+type MatchRecursive struct{}
+
+type Literal string
+type Identifier string
+
+// Variable represents a variable in the path template with its field path and nested segments.
+type Variable struct {
+	FieldPath []*Identifier
+	Segments  []*Segment
+}
+
+// Segment represents a single segment of the path template, which can hold one of several types of values.
+type Segment struct {
+	Literal        *Literal
+	Match          *Match
+	MatchRecursive *MatchRecursive
+	Variable       *Variable
+}
+
+const eof = -1
+
+const (
```

Comment:
merge?

Reply:
was keeping them separate because they represent diff things, but it was looking kinda odd to have two diff const declarations side by side. Merged